### PR TITLE
feat: [CO-699] add cleanup script

### PR DESCRIPTION
**Pull request by frisonisland**
# Goal
Add cleanup script
See also: 
- https://github.com/Zextras/carbonio-build/pull/58
- https://bitbucket.org/zextras/carbonio-core/pull-requests/46

# How
The script behaves similarly to ldapattributeupdate, except:
- It checks current version and executes cleanup scripts in order of timestamp from cleanup directory.
- It updates schema version after each file execution
- it calls exit(1) in case of failure (however the postinstall does not fail and proceeds)
- it does not print not eligible cleanups
- Prints "Current installed LDAP" instead of "Installed LDAP"
- Adds a separator at beginning, end, and after each file instruction is processed.

# Example

I made simple examples with a deprecated attribute.

## Manual run
```--------------------------------------------------------------------------------
Starting LDAP cleanup...
----------------------------------------
Current installed LDAP Schema Version: 2681988720 
--------------------------------------------------------------------------------
Executing cleanup from 2681988723.json
Found 0 entries with attribute zimbraFeatureTouchClientEnabled.
LDAP schema upgraded to version 2681988723 
--------------------------------------------------------------------------------
Executing cleanup from 2681988724.json
Found 0 entries with attribute zimbraFeatureTouchClientEnabled.
LDAP schema upgraded to version 2681988724 
--------------------------------------------------------------------------------
Executing cleanup from 2681988725.json
Found 0 entries with attribute zimbraFeatureTouchClientEnabled.
LDAP schema upgraded to version 2681988725 
--------------------------------------------------------------------------------
Cleanup completed.
--------------------------------------------------------------------------------

```

## Postinstall of carbonio-directory-server
```
Setting up carbonio-directory-server (4.0.22-1ubuntu1~focal) ...
*Starting LDAP upgrade...
--------------------------------------------------------------------------------
Starting LDAP cleanup...
----------------------------------------
Current installed LDAP Schema Version: 2681988720 
LDAP cleanup directory(/opt/zextras/common/etc/openldap/zimbra/cleanup/attrs) not found.
Unable to process LDAP cleanup.
Killing slapd with pid 1725927. done.
Looking for LDAP installation...succeeded
Installing core schema...
Installing cosine schema...
Installing inetOrgPerson schema...
Installing zimbra schema...
Installing amavis schema...
Installing dyngroup schema...
Installing OpenDKIM schema...
----------------------------------------
Installed LDAP Schema Version: 2681988720 
--------------------------------------------------------------------------------
Skipping: 1645791607.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1649167975.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1656290447.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1656504433.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1665746833.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1670857065.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1674157658.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1677585425.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1680172610.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
Skipping: 1680512235.json
    Reason: not eligible for this update.
--------------------------------------------------------------------------------
```

### CODE REVIEW

File: src/libexec/ldapattributecleanup [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+my $ldap_attribute_cleanup_dir = "$source_config_dir/zimbra/cleanup/attrs";
+my $ldap_root_password = getLocalConfig("ldap_root_password");
+my $ldap_master_url = getLocalConfig("ldap_master_url");
+my $ldap_is_master = getLocalConfig("ldap_is_master");
+my $ldap_starttls_supported = getLocalConfig("ldap_starttls_supported");
+my $zimbra_tmp_directory = getLocalConfig("zimbra_tmp_directory");
+
+if (lc($ldap_is_master) ne "true") {
+    exit(1);
+}
```
Comment (Polpetta) [inline src/libexec/ldapattributecleanup:46]: Does it make sense to put a simple error description to stderr before exiting?

File: src/libexec/ldapattributecleanup [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            capath => "/opt/zextras/conf/ca",
+        ) or die "start_tls: $@";
+        $mesg->code && die "TLS: " . $mesg->error . "\n";
+    }
+}
+
+# bind ldap or exit with error on failure fail
+$mesg = $ldap->bind("cn=config", password => "$ldap_root_password");
+if ($mesg->code()) {
+    print "Unable to bind: $!";
```
Comment (Polpetta) [inline src/libexec/ldapattributecleanup:73]: If this is an error message it should be logged to `stderr` instead of `stdout`

**Review by Polpetta — CHANGES_REQUESTED** [review]
Logic seems ok, since it is new codebase I'd start to print errors into `stderr` in order to properly use linux output streams

**Comment by frisonisland** [comment]
> LGTM, what about the backup step? Will it be asked separately as step in upgrade notes for the upcoming release?

Yes, we decided the backup is out of scope for now, so backup is manually done.

**Review by keshavbhatt — APPROVED** [review]
LGTM.

File: src/libexec/ldapattributecleanup (hunk 1 above)
Comment (frisonisland) [reply src/libexec/ldapattributecleanup]: I added a STDERR message

File: src/libexec/ldapattributecleanup (hunk 2 above)
Comment (frisonisland) [reply src/libexec/ldapattributecleanup]: Added stderr

**Comment by frisonisland** [comment]
@keshavbhatt @Polpetta I added a return status to apply_cleanup() and update_zimbra_ldap_schema_version() ro returns a boolean representing failure status. I think it needs further refinement to avoid leaving ldap connection open, etc.

**Review by keshavbhatt — APPROVED** [review]
LGTM.

File: src/libexec/ldapattributecleanup [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    my ($attributes) = @_;
+    foreach my $attribute (@$attributes) {
+        $mesg = $ldap ->search(
+            base=>"",
+            filter=>"&(objectClass=*)($attribute=*)",
+            scope=>"sub",
+        );
+        my @entries = $mesg->entries;
+        my $size = @entries;
+        print STDOUT "Found $size entries with attribute $attribute.\n";
```
Comment (keshavbhatt) [inline src/libexec/ldapattributecleanup:237]:
```suggestion
        print STDOUT "Found $size entries with attribute $attribute\n";
```
I was testing this and found that `.` after attributes names ugly

```Executing cleanup from 1681988810.json
Found 0 entries with attribute dcarbonioAdminPort.
Found 0 entries with attribute dcarbonioAdminPortTwo.
Found 0 entries with attribute dcarbonioAdminPortThree.
```

Comment (frisonisland) [reply src/libexec/ldapattributecleanup:237]: Thx, I was checking what was wrong with these attribute names at first :laughing:

**Review by Polpetta — APPROVED** [review]

**Pull request merged**